Fix Null dereference

JohnnySavages opened the pull request:
Returning NULL from sk_X509_value is possible if num_trusted is less than 0 or if one of the ctx->chain->data[i] values is NULL.

Found by Linux Verification Center (linuxtesting.org) with SVACE.

Refer: https://github.com/openssl/openssl/pull/23273

fwh-dc reviewed (approved):
I think it is okay to add the guard even though I am uncertain if the failing condition is ever met.

nhorman commented on crypto/x509/x509_vfy.c:
```diff
@@ -1059,6 +1059,8 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
     for (i = num_untrusted; i < num; i++) {
         x = sk_X509_value(ctx->chain, i);
+        if (x == NULL)
+            return -1;
```
num_untrusted is used higher up in this function:
```
/*
     * Check for a DANE issuer at depth 1 or greater, if it is a DANE-TA(2)
     * match, we're done, otherwise we'll merely record the match depth.
     */
    if (DANETLS_HAS_TA(dane) && num_untrusted > 0 && num_untrusted < num) {
        trust = check_dane_issuer(ctx, num_untrusted);
        if (trust != X509_TRUST_UNTRUSTED)
            return trust;
    }
```
I haven't checked but I imagine check_dane_issuer will have similar problems to this code .

It might be better to do a num_untrusted < 0 sanity check at the top of this function in addition to this null check.

JohnnySavages replied:
check_dane_issuer sk_X509_value result being checked: https://github.com/openssl/openssl/blob/d31fbbeded371bf30695727841b2e8da855c7ffa/crypto/x509/x509_vfy.c#L3421

JohnnySavages replied:
@nhorman

npajkovsky replied:
It's quite a questionable fix. The chain should not contain NULL gaps; returning NULL is impossible.

If we want to fix it, the return value should be X509_TRUST_UNTRUSTED.

There is another occurrence in the same function that does not check the return value, but also, it should be impossible to return NULL 

```c
        i = 0;
        x = sk_X509_value(ctx->chain, i);
```

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

JohnnySavages committed:
Fix Null dereference

Returning NULL from sk_X509_value is possible if num_trusted is less than 0 or if one of the ctx->chain->data[i] values is NULL.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago
